### PR TITLE
P1009R2 Array size deduction in new-expressions

### LINKED ISSUE #2690: [2019-02 CWG Motion 10] P1009R2 Array size deduction in new-expressions
[P1009R2](https://wg21.link/p1009r2)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #2690.

### CODE REVIEW

**Comment by burblebee** [comment]
Review completed - see comments below:
* The paper directs that the new text at the end of paragraph 6 should be part of paragraph 6, while the changes in the PR adds it as a new paragraph.

File: source/expressions.tex [hunk 1]
```diff
@@ -4532,6 +4532,15 @@
 constant expression).
 \end{example}
 
+\pnum
```
Comment (burblebee) [inline source/expressions.tex]: Should this be part of paragraph 6 as directed by the paper?

Comment (jensmaurer) [reply source/expressions.tex]: First, the editing instructions in the incoming paper do say "change p6", but the actual layout of the paper text suggests that the text should actually be in a new paragraph.
Second, p6 (actually p7, it seems) talks about (really-present) _constant-expression_s and _expression_s in the _new-type-id_, whereas the new sentence talks about the absence of an expression.
So, actually I think the right fix is to move the new text slightly earlier, before we start talking about _expression_.

Comment (jensmaurer) [reply source/expressions.tex]: Add-on patch (totally optional) pushed; @burblebee, what do you think?

Comment (burblebee) [reply source/expressions.tex]: I like it :).

File: source/expressions.tex [hunk 2]
```diff
@@ -4542,9 +4539,11 @@
 the \grammarterm{new-initializer}~(\ref{dcl.init.aggr}, \ref{dcl.init.string}).
 
 \pnum
+\indextext{\idxcode{new}}%
+If the \grammarterm{expression} in a \grammarterm{noptr-new-declarator}
+is present, it is implicitly converted to \tcode{std::size_t}.
 \indextext{function!allocation}%
```
Comment (burblebee) [inline source/expressions.tex]: Do you think this index should precede the new text?

Comment (jensmaurer) [reply source/expressions.tex]: It's existing stuff, and I'm not too happy about the indexing here in general. I'd like to leave it as-is for now.

**Pull request merged**